src/entities/shooter.cpp

```diff
@@ -0,0 +1,117 @@
+            shootCoolDown_ = 0;
+        }
+    }
+
+    bool Shooter::canShoot(){
+        return shootCoolDown_ == 0;
+    }
+    
+    void Shooter::attack(manager::Level *levelManager) {
+        if (!canShoot()) {
+            this->ShootCoolDown();
+            return;
+        }
+        level::Level *level = levelManager->getLevel();
+        Position playerPosition = levelManager->getPlayer()->getPosition();
+        Position currPosition = this->getPosition();
+        
+        // TODO simo, c'è ancora bisogno di migliorare questa parte
+        if(currPosition.riga == playerPosition.riga){
+            if(currPosition.colonna < playerPosition.colonna){
+                if(LineFirable(currPosition.riga, currPosition.colonna + 1, playerPosition.colonna, levelManager)){
+                    this->resetShootCoolDown();
+                    logger_.info("enemy firing a bullet");
+                    Position bulletPosition = Position(currPosition.riga, currPosition.colonna + 1);
+                    weapon::Bullet *bullet = new weapon::Bullet(bulletPosition, enums::Direction::RIGHT, this->getAttack());
+                    level->addBullet(bullet);
+                }
+            }
+            else{
+                if(LineFirable(currPosition.riga, playerPosition.colonna + 1, currPosition.colonna, levelManager)){
+                    this->resetShootCoolDown();
+                    logger_.info("enemy firing a bullet");
+                    Position bulletPosition = Position(currPosition.riga, currPosition.colonna - 1);
+                    weapon::Bullet *bullet = new weapon::Bullet(bulletPosition, enums::Direction::LEFT, this->getAttack());
+                    level->addBullet(bullet);
+                }
+            }
+        }
+        if(currPosition.colonna == playerPosition.colonna){
+            if(currPosition.riga < playerPosition.riga){
+                if(columnFirable(currPosition.colonna, currPosition.riga + 1, playerPosition.riga, levelManager)){
+                    this->resetShootCoolDown();
+                    logger_.info("enemy firing a bullet");
+                    Position bulletPosition = Position(currPosition.riga + 1, currPosition.colonna);
+                    weapon::Bullet *bullet = new weapon::Bullet(bulletPosition, enums::Direction::DOWN, this->getAttack());
+                    level->addBullet(bullet);
+                }
+            }
+            else{
+                if(columnFirable(currPosition.colonna, playerPosition.riga + 1, currPosition.riga, levelManager)){
+                    this->resetShootCoolDown();
+                    logger_.info("enemy firing a bullet");
+                    Position bulletPosition = Position(currPosition.riga - 1, currPosition.colonna);
+                    weapon::Bullet *bullet = new weapon::Bullet(bulletPosition, enums::Direction::UP, this->getAttack());
+                    level->addBullet(bullet);
+                }
+            }
+        }
+    }
```

Comment:
Ho guardato simo, si può ancora fare meglio, scusa se non ho spiegato bene quello che volevo ci fosse per il canAttack.
In pratica non hai bisogno di tutti quei lifeOfFire.
La funzione canAttack ti dovrebbe dire se puoi sparare o meno (in modo molto simile ai lineofFire attuali).
Una volta che conosci questa cosa, se è false smetti.
Altrimenti ti calcoli la direzione per sparare
Così non devi più fare nessun check ripetuto 4 volte.
```cpp
if (this->canAttack(levelManager)) {
    Diretion dir = find-direction-to-player;
    Istantiate bullet with correct direction
    fire the bullet 
} else return;
```